### ISSUE TITLE
Freezing behavior when showing `AbstractArray`s after `using StaticArrays`

### ISSUE THREAD

**Issue by andyferris** [JuliaLang/julia#20908]
I'm getting some odd behavior lately with my rewrite of *StaticArrays*. It is possible that it is my mistake with writing the package - but the result seems (to me at least) to be compiler related, and the behavior happened after I rebuilt Julia (I haven't attempted to bisect as of yet, sorry).

The short story is, if you try and run the (work-in-progress) `julia-0.6` branch of *StaticArrays* on master, then I see freezes whenever you try and print an `AbstractArray`. Oddly, these freezes *aren't* happening for `StaticArray`s, but only for `Vector` and so-on. In the below you can see `^C` where I got impatient and pressed CTRL-C.  However, these freezes only happen appear to happen when preparing a specialization for the first time since `using StaticArrays` - the array can print fine without error, but it locks up in the first instance (for each eltype of `Vector`, say).

```julia
              _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-pre.alpha.55 (2017-03-06 07:55 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 73e20a3 (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> using StaticArrays
INFO: Recompiling stale cache file /home/ferris/.julia/lib/v0.6/StaticArrays.ji for module StaticArrays.

julia> [1,2,3]
^C^C^C^C^C^CWARNING: Force throwing a SIGINT
^C3-element Array{Int64,1}:
 1
 2
 3

julia> [1,2,3]
3-element Array{Int64,1}:
 1
 2
 3

julia> [1.0,2.0,3.0]
^C^C^C^C^C^CWARNING: Force throwing a SIGINT
^C3-element Array{Float64,1}:
^CError showing value of type Array{Float64,1}:
ERROR: InterruptException:
Stacktrace:
 [1] #showarray#254(::Bool, ::Function, ::IOContext{Base.Terminals.TTYTerminal}, ::Array{Float64,1}, ::Bool) at ./show.jl:1697
 [2] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::Array{Float64,1}) at ./REPL.jl:122
 [3] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::Array{Float64,1}) at ./REPL.jl:125
 [4] display(::Array{Float64,1}) at ./multimedia.jl:194

julia> SVector(1,2,3)
3-element StaticArrays.SVector{3,Int64}:
 1
 2
 3
```

(as an aside, this was an "interesting" exercise in debugging - the earlier in my code I tried to debug the problem (using `println` or `@show` or the REPL), the earlier I got the freeze!)

cc: @c42f

**Comment by timholy** [comment]
Have you tried building Julia with #20874?

**Comment by andyferris** [comment]
I can try - but it's not 20 seconds, I think I waited 10 minutes once.

**Comment by andyferris** [comment]
It tuned out that I had overloaded a method which intersected with `getindex(a::AbstractArray, inds::Int...)` (it was really `getindex(a::AbstractArray, inds::Union{Int, StaticArray{Int}}...)`) which I guessed caused a fair bit of recompilation. Making this signature in *StaticArrays* more sensible removes my problem.

However, it's still not clear why this freeze had occurred... defining `@Base.propagate_inbounds Base.getindex(a::AbstractArray, ind::Int...) = a[sub2ind(size(a), ind...)]` at the REPL doesn't cause any significant slowdown. It was long enough that I was worried that it would never terminate.
